memgpt/local_llm/chat_completion_proxy.py

```diff
@@ -0,0 +1,88 @@
+"""MemGPT sends a ChatCompletion request
+
+Under the hood, we use the functions argument to turn
+"""
+
+
+"""Key idea: create drop-in replacement for agent's ChatCompletion call that runs on an OpenLLM backend"""
+
+import os
+import json
+import requests
+
+from .webui_settings import DETERMINISTIC, SIMPLE
+from .llm_chat_completion_wrappers import airoboros
+
+HOST = os.getenv('OPENAI_API_BASE')
+HOST_TYPE = os.getenv('BACKEND_TYPE')  # default None == ChatCompletion
+
+
+class DotDict(dict):
+    """Allow dot access on properties similar to OpenAI response object"""
+
+    def __getattr__(self, attr):
+        return self.get(attr)
+
+    def __setattr__(self, key, value):
+        self[key] = value
+
+
+async def get_chat_completion(
+        model,  # no model, since the model is fixed to whatever you set in your own backend
+        messages,
+        functions,
+        function_call="auto",
+    ):
+    if function_call != "auto":
+        raise ValueError(f"function_call == {function_call} not supported (auto only)")
+
+    if True or model == 'airoboros_v2.1':
+        llm_wrapper = airoboros.Airoboros21Wrapper()
+
+    # First step: turn the message sequence into a prompt that the model expects
+    prompt = llm_wrapper.chat_completion_to_prompt(messages, functions)
+    # print(prompt)
+
+    if HOST_TYPE != 'webui':

```

Comment:
This is presumably where you would then call to different functions depending on the web server you're running, since they'll have different api addresses / ports (not all just `v1/generate/`